### ISSUE TITLE
pipx fails on some packages where pipsi works

### ISSUE THREAD

**Issue by apnewberry** [pipxproject/pipx#115]
pipx doesn't find the binary for this package but pipsi does.

```
$ pipx install py-spy
No binaries associated with package py-spy. 
$ pipx --version
0.12.3.3
$ pipsi --version
pipsi, version 0.9, python /usr/bin/python3

$ pipsi install py-spy
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/user/.local/venvs/py-spy/bin/python3
Also creating executable in /home/user/.local/venvs/py-spy/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
Collecting py-spy
  Using cached https://files.pythonhosted.org/packages/26/c2/bacd6ff83b43e1eff51d0d93a1fc930daad02fa93c0499d6facece57f786/py_spy-0.1.10-py2.py3-none-manylinux1_x86_64.whl
Installing collected packages: py-spy
Successfully installed py-spy-0.1.10
  Linked script /home/user/.local/bin/py-spy
Done.

```

**Comment by cs01** [comment]
Thanks for sharing. If anyone has a chance to look into this it would be appreciated. 

pipsi's code for doing this is located here:
https://github.com/mitsuhiko/pipsi/blob/master/pipsi/scripts/find_scripts.py

and pipx's is here:
https://github.com/pipxproject/pipx/blob/master/pipx/venv_metadata_inspector.py

pipx and pipsi call into this modules via subprocess calls, so they can be run directly to debug and inspect how they work.

**Comment by amcgregor** [comment]
The proliferation of these custom bindings of individual tools to virtual environment awareness, and notably, half dozen different approaches to making a venv-aware pip, I honestly do not grok why this seems to be a desired thing.  Why does this seem to be a good idea vs. "proper" use of virtual environments, directly?

(Apologies if this animated SVG terminal capture is gibbled; waving mouse over it seems to trigger proper rendering; also clickable.)

[![just-use-venv](http://s.webcore.io/7ee6d99428c8/115.svg)](http://s.webcore.io/7ee6d99428c8/115.svg)

This uses a tiny, tiny little [venv shell script](https://github.com/amcgregor/dotfiles/blob/dotfiles/.zsh/func/venv) ('autoloaded' in the cast) to patch the `activate` script to expose the actual root path of the project (one directory up) instead of the internal hidden `.venv` as `$VIRTUAL_ROOT`, as used by PS1 abbreviation.  Then, any time the CWD changes, [the `chpwd()` function is invoked](https://github.com/amcgregor/dotfiles/blob/dotfiles/.zsh/hook/zshrc/functions.zsh#L21), which checks to see if we have entered a new, or exited an active virtual environment root path, and does the right thing.  (Activate, deactivate, etc.)

Noting that with this there are at least three different approaches to use possible:

* Virtual environment as project — slap a `src` directory inside `$VIRTUAL_ENV`, go from there.
* Hidden virtual environment within project — the script above uses this, populating a `.venv` directory within your project folder.
* Symlink that, son!  You can freely move that .venv directory anywhere you want, and symlink it back into place from there.  E.g. I have a bunch of environments in `~/Library/Environments/`.

**Comment by cs01** [comment]
@apnewberry I just tried to reproduce this with 0.13.1.1 and it worked fine:
```
>> pipx install py-spy
  installed package py-spy 0.1.10, Python 3.6.7
  These binaries are now globally available
    - py-spy
done! ✨ 🌟 ✨
/home/csmith/git csmith@gram
>> py-spy -h
py-spy 0.1.10
A sampling profiler for Python programs

USAGE:
    py-spy [FLAGS] [OPTIONS] --pid <pid> [python_program]...

FLAGS:
        --dump           Dump the current stack traces to stdout
    -F, --function       Aggregate samples by function name instead of by line number
    -h, --help           Prints help information
        --nonblocking    Don't pause the python process when collecting samples. Setting this option will reduce the
                         perfomance impact of sampling, but may lead to inaccurate results
    -V, --version        Prints version information

OPTIONS:
    -d, --duration <duration>    The number of seconds to sample for when generating a flame graph [default: 2]
    -f, --flame <flamefile>      Generate a flame graph and write to a file
    -p, --pid <pid>              PID of a running python program to spy on
    -r, --rate <rate>            The number of samples to collect per second [default: 100]

ARGS:
    <python_program>...    commandline of a python program to run

```

**Comment by cs01** [comment]
@amcgregor nice! I am a big fan of venv and I use a similar shell function I called `venvhere` which I think does the same thing as your `venv` function :
```
venvhere () 
{ 
    python3 -m venv venv --prompt $(basename $PWD) && source venv/bin/activate && pip install -U pip --quiet
}
```

> Why does this seem to be a good idea vs. "proper" use of virtual environments, directly?

It is only a good idea *for certain workflows*. I **agree** that creating and manually managing venv's for development of a project is the best way to do it. pipx is definitely not meant to be a replacement for all venv interactions. 

Its goal is to make CLI binaries accessible globally, that's it. Using venv is the best way for pipx to do that. pipx saves you a few steps, and makes a single UI to list, upgrade, uninstall, these binaries/packages. I actually don't know of an easy way to know which entry points a package offers other than using pipx or inspecting the package's source code.

If you don't find value in it, that is totally fine! No need to use it. I made it to solve a problem: simplify and improve the installation process of another [Python program](gdbgui.com) I made.
